File: neetCode150/py/543DiameterofBinaryTree.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def diameterOfBinaryTree(self, root):
        diameter = 0
        def dfs(root):
            nonlocal diameter
            if not root: 
                return 0
            left = dfs(root.left)
            right = dfs(root.right)

            diameter = max(diameter, left+right)
            return 1+ max(left,right)
        
        dfs(root)
        return diameter

root = TreeNode(1)
root.left = TreeNode(2)
root.right = TreeNode(3)
root.left.left = TreeNode(4)
root.left.right = TreeNode(5)

sol = Solution()
result = sol.diameterOfBinaryTree(root)
print(result)

# Time Complexity = O(n)
# Space Complexity = O(h)

File: 2025/Q4LeetCode/543_DiameterofBinaryTree.js
// 543. Diameter of Binary Tree

const diameterOfBinaryTree = (root) => {
  let diameter = 0;

  const depth = (node) => {
    if (!node) return 0;
    const left = depth(node.left);
    const right = depth(node.right);
    diameter = Math.max(diameter, left + right);
    return Math.max(left, right) + 1;
  };

  depth(root);

  return diameter;
};

let root = [1, 2, 3, 4, 5];
let tree = {};

const buildTree = (arr, index = 0) => {
  if (index >= arr.length || arr[index] === null) return null;
  const node = { val: arr[index], left: null, right: null };
  node.left = buildTree(arr, 2 * index + 1);
  node.right = buildTree(arr, 2 * index + 2);
  return node;
};

root = buildTree(root);
console.log(diameterOfBinaryTree(root)); // Output: 3
